crypto/merkle/src/lib.rs

```diff
@@ -0,0 +1,26 @@
+use hashes::{Hash, Hasher, MerkleBranchHash, ZERO_HASH};
+
+pub fn calc_merkle_root(hashes: impl ExactSizeIterator<Item = Hash>) -> Hash {
+    let next_pot = hashes.len().next_power_of_two();
+    let vec_len = 2 * next_pot - 1;
+    let mut merkles = vec![None; vec_len];
+    for (i, hash) in hashes.enumerate() {
+        merkles[i] = Some(hash);
+    }
+    let mut offset = next_pot;
+    for i in (0..vec_len - 1).step_by(2) {
+        if merkles[i].is_none() {
+            merkles[offset] = None;
+        } else {
+            merkles[offset] = Some(merkle_hash(merkles[i].unwrap(), merkles[i + 1].unwrap_or(ZERO_HASH)));
+        }
+        offset += 1
+    }
+    merkles.last().unwrap().unwrap()
```

Comment:
FYI, this can be done much more efficiently without a vector, by iterating through the tree, and hashing what's needed

Reply:
(Without allocating any vector)